Contracts Cleanup (#25)
stlab/better-code: better-code/src/chapter-2-contracts.md

Hoare used this notation, called a “Hoare triple,”

> > $\{P\}S\{Q\}$

Reviewer: As I tried to suggest elsewhere, maybe we should be postponing any “isn't pretty” changes in favor of whatever makes writing fastest and easiest to understand when reviewing.  I agree we should be using LaTeX or some similar syntax for math eventually.  I just am not convinced it's a good use of time at this point to spend time changing something *N²* into some more complex form, especially if it won't easily render everywhere correctly.

which is an assertion that if **precondition** $P$ is met, operation
$S$ establishes **postcondition** $Q$.

> <!-- This had been using math for pre and post conditions, but I find mixing math and code makes it look like we are talking about different `x` and $x$ variables and equality vs. assignment gets confusing. I think if the operation is expressed in code, the conditions should be expressed in code. -->

Reviewer: @sean-parent WDYT?

> <!-- This had been using math for pre and post conditions, but I find mixing math and code makes it look like we are talking about different `x` and $x$ variables and equality vs. assignment gets confusing. I think if the operation is expressed in code, the conditions should be expressed in code. -->

Reviewer: Right, it's messy and friction-inducing.  Why do this now?

For example:

- if we start with `x == 2` (precondition), after `x += 1`, `x == 3` (postcondition):

  > $\lbrace$ `x == 2` $\rbrace$ `x += 1` $\lbrace$ `x == 3` $\rbrace$

- if `x` is less than the maximum integer (precondition), after `x
  += 1`, `x` is greater than the minimum integer (postcondition):

  > $\lbrace$ `x < Int.max` $\rbrace$ `x += 1` $\lbrace$ `x > Int.min` $\rbrace$

What makes preconditions and postconditions useful for formal proofs
is this *sequencing rule*:

> > $\{P\}S\{Q\} \wedge \{Q\}T\{R\} \Rightarrow \{P\}S;T\{R\}$

Reviewer: We definitely want good math output for final production, but for the purposes of just reading and reviewing the text and changes, I really don't like the engineering cost  I can't easily verify that it means what I originally wrote (and as noted elsewhere GitHub doesn't render the curlies so can't verify that way).

Author: Fixed the curlies here too (push pending) - but verbose.

> All undocumented software is waste. It's a liability for a company.
>

> > &mdash; Alexander Stepanov (<https://youtu.be/COuHLky7E2Q?t=1773>)

Reviewer: You added a space after the dash.  I don't think that's how this is done ([non-authoritative citation](https://www.wisdomandwonder.com/writing/13270/what-dash-should-i-use-when-attributing-a-quotation)).

```suggestion
> &mdash;Alexander Stepanov (<https://youtu.be/COuHLky7E2Q?t=1773>)
```

explicit precondition that isn't implied by the summary: it requires
that the predicate be a strict weak ordering.

> <!-- SRP: this section bothers me. "among others" instead of fully spelling out the requirements, using (i, j + 1) which may not exist, and the n^2 comparisons without calling out the O(n^3) complexity or which properties could be practically checked. Also is "stable" the term we want to use? Regular and deterministic are also candidates. I've tried to rewrite this a couple of times, but it just gets too complex and the main point is lost. -->

Reviewer: I think you may have some valid points here but too many complaints and a little to diffuse to deal with here; let's chat in person.

Reviewer: I do think "deterministic" is a better word than "stable".

_Some_ precondition on the predicate is needed just to make the result
a meaningful sort with respect to the predicate.  For example, a
totally unconstrained predicate could return random boolean values,
and there's no reasonable sense in which the function could be said to
leave the elements sorted with respect to that.  Therefore the

> predicate at least has to be stable. To leave elements meaningfully

Reviewer: ```suggestion
predicate at least has to be deterministic. To leave elements meaningfully
```

function's guarantees under one name: its postconditions.

The strict weak ordering requirement is a great example of a

> precondition that can't be efficiently checked.  Even if we could assume the
> comparison is stable, to do so would
> require at least $n^2$ comparisons, where $n$ is the number of

Reviewer: ```suggestion
precondition that can't be checked: there's no way to verify that a function
is deterministic. Even if we could assume determinism, a complete check
requires at least $n^3$ comparisons, where $n$ is the number of
```

probably put a link in the documentation to a definition.

```swift

>   /// - Precondition: `areInIncreasingOrder` is [a strict weak

Reviewer: What's the rationale for these spaces?  I don't see how they're helpful.  Personally I think they look quite awkward in the rendered output.

Reviewer: OK, it's consistent with the rest of the chapter (except for the several(!) code blocks where indentation is totally messed up).  But maybe we should reconsider and strip the leading indentation from code blocks.  Horizontal space is at a premium in print books (do we still believe in those?) and anyway to my eye, isolated indentation looks weird on a gray backgound and does nothing useful on a white background.
